include/opentracing/propagation.h

```diff
@@ -89,6 +98,18 @@ class TextMapReader {
  public:
   virtual ~TextMapReader() = default;
 
+  // LookupKey returns the value for the specified `key` if available. If no
+  // such key is present, it returns `key_not_found_error`.
+  //
+  // TextMapReaders are not required to implement this method. If not supported,
+  // the function returns `lookup_key_not_supported_error`.
+  //
+  // Tracers may use this as an alternative to `ForeachKey` as a faster way to
+  // extract span context.
+  virtual expected<string_view> LookupKey(string_view /*key*/) const {
```

Comment:
I'm super rusty in C++, but doesn't requiring the tracer implementation to run in try-catch introduces its own overhead? Exceptions should be used for exceptional situations, it's not an error or exception for the carrier to not support by-key lookup. I'd rather have something like `optional<KeyedTextMapReader> getKeyedReader()`

Reply:
This is an error code approach: it doesn't use exceptions.

In some of the projects that might use this (e.g. envoy, nginx), they index some of the keys but not all of them; so I tried to design it in a way that key-value lookup is supported on a per-key basis, not just a simple on/off for all keys.

Reply:
Wouldn't it be better to have this as an additional interface, which implementations can inherit? e.g. 
    class Indexable
    {
     public:
         virtual optional<string_view> operator[](string_view) const = 0;
    };

This way if you want to test if it is supported you can do a dynamic_cast. If you are some how certain that it will be supported you can just do a static_cast(I think) to avoid any RTTI overhead. You don't have to handle the not supported case in every call to LookupKey either, which will have its own overhead, even if it is just instruction cache use or branch prediction.

Reply:
@pantoss Lookup needs to be supported to be on a per-key level. Some applications don't index all of the header keys. For example, envoy provides fast lookup only for a list of [hard-coded keys](
https://github.com/envoyproxy/envoy/blob/master/source/common/http/header_map_impl.h#L102).